### PR TITLE
Support disrtributed hive cluster query

### PR DESCRIPTION
**Pull request by lgbo-ustc**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add storage engine `HiveCluster` and table funtion `hiveCluster` to support distribted query for hive.

query by table funtction `hiveCluster`
```sql
SELECT 
    day, count(*) 
FROM hiveCluster(
    'cluster_simple', -- clickhouse cluster
    'thrift://hivetest:9083', -- hive metastore url
    'tmp', -- hive table's database
    'demo', -- hive table's name
    'id Nullable(String), score Nullable(Int32), day String',  -- table structure
    'day' -- partition keys, comma separated
) 
GROUP BY day ORDER BY day
```

query by storage `HiveCluster`
```sql
CREATE TABLE IF NOT EXISTS default.demo_parquet ON CLUSTER cluster_simple (
    `id` Nullable(String), `score` Nullable(Int32), `day` String -- table structure
) 
ENGINE = HiveCluster(
    'cluster_simple', -- clickhouse cluster
    'thrift://hivetest:9083', -- hive metastore url
    'test', -- hive table's databse
    'demo' -- hive table's name
)
PARTITION BY(day);

SELECT day, count(*) FROM default.demo_parquet GROUP BY day ORDER BY day
```

Some big changes to notice
1. `StorageHiveCluster` is based on `StorageHive`, and `StorageHive` has a little big change. Move the hive files collect action in `StorageHive` into a single class `LocalHiveSourceFilesCollector`

Some benchmark results
1. run group by query by using storage engine `Hive` on single node
```
11 rows in set. Elapsed: 7.728 sec. Processed 119.83 million rows, 24.63 GB (15.51 million rows/s., 3.19 GB/s.

```
2. run group by query by using storage engine `HiveCluster` on three nodes cluster
```
11 rows in set. Elapsed: 3.745 sec. Processed 119.83 million rows, 24.63 GB (32.00 million rows/s., 6.58 GB/s.)
```

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by lgbo-ustc** [comment]
@nikitamikhaylov Since it a long time for no feedback. May I ask how is it going on

File: src/Storages/StorageS3Cluster.h [hunk 1]
```diff
@@ -25,6 +25,8 @@ class StorageS3Cluster : public shared_ptr_helper<StorageS3Cluster>, public ISto
 public:
     std::string getName() const override { return "S3Cluster"; }
 
+    bool isRemote() const override { return true; }
```
Comment (zhanglistar) [inline src/Storages/StorageS3Cluster.h:28]: del

File: src/Storages/HDFS/StorageHDFSCluster.h [hunk 2]
```diff
@@ -24,6 +24,8 @@ class StorageHDFSCluster : public shared_ptr_helper<StorageHDFSCluster>, public
 public:
     std::string getName() const override { return "HDFSCluster"; }
 
+    bool isRemote() const override { return true; }
```
Comment (zhanglistar) [inline src/Storages/HDFS/StorageHDFSCluster.h:27]: del

File: src/IO/Marshallable.h [hunk 3]
```diff
@@ -0,0 +1,488 @@
+#pragma once
```
Comment (zhanglistar) [inline src/IO/Marshallable.h:1]: use json

File: src/Processors/Formats/IRowInputFormat.h [hunk 4]
```diff
@@ -44,6 +44,11 @@ class IRowInputFormat : public IInputFormat
 
     Chunk generate() override;
 
+    void onCancel() override
```
Comment (zhanglistar) [inline src/Processors/Formats/IRowInputFormat.h:47]: del

**Review by zhanglistar — COMMENTED** [review]
1. use json instead of marshal
2. sql simplify for easy use

**Comment by lgbo-ustc** [comment]
@nikitamikhaylov Do you have time to take a review on this ?

**Comment by tavplubix** [comment]
All changes related to hive are blocked by https://github.com/ClickHouse/ClickHouse/issues/43541 (the tests are temporarily disabled, so we cannot merge new features)

**Pull request closed without merging**